toolchain/check/function.cpp

```diff
@@ -191,60 +191,88 @@ auto CheckFunctionTypeMatches(Context& context,
                      context.functions().Get(prev_function_id), substitutions);
 }
 
+// Checks to see if a structurally valid redeclaration is allowed in context.
+// These all still merge.
+static auto CheckIsAllowedRedecl(Context& context, Parse::NodeId node_id,
+                                 SemIR::Function& new_function,
+                                 bool new_is_definition,
+                                 SemIR::Function& prev_function,
+                                 bool prev_is_import) -> void {
+  CARBON_DIAGNOSTIC(FunctionPreviousDecl, Note, "Previously declared here.");
+  if (prev_is_import) {
+    // TODO: Allow non-extern declarations in the same library.
+    if (!new_function.is_extern && !prev_function.is_extern) {
+      CARBON_DIAGNOSTIC(
+          FunctionNonExternRedecl, Error,
+          "Only one library can declare function {0} without `extern`.",
+          SemIR::NameId);
+      context.emitter()
+          .Build(node_id, FunctionNonExternRedecl, prev_function.name_id)
+          .Note(prev_function.decl_id, FunctionPreviousDecl)
+          .Emit();
+      return;
+    }
+  } else {
+    if (!new_is_definition) {
+      CARBON_DIAGNOSTIC(FunctionRedecl, Error,
+                        "Redundant redeclaration of function {0}.",
+                        SemIR::NameId);
+      context.emitter()
+          .Build(node_id, FunctionRedecl, prev_function.name_id)
+          .Note(prev_function.decl_id, FunctionPreviousDecl)
+          .Emit();
+      return;
+    }
+    if (prev_function.definition_id.is_valid()) {
+      CARBON_DIAGNOSTIC(FunctionRedefinition, Error,
+                        "Redefinition of function {0}.", SemIR::NameId);
+      CARBON_DIAGNOSTIC(FunctionPreviousDefinition, Note,
+                        "Previously defined here.");
+      context.emitter()
+          .Build(node_id, FunctionRedefinition, prev_function.name_id)
+          .Note(prev_function.definition_id, FunctionPreviousDefinition)
+          .Emit();
+      return;
+    }
+    if (new_is_definition && prev_function.is_extern) {
```

Comment:
I think we're missing a diagnostic for:
```
fn F();
extern fn F() {}
```
I wonder if perhaps we could split this up as:

1) Diagnose if `prev_function.is_extern != new_function.is_extern`.
2) Diagnose if `new_is_definition && new_function.is_extern`.

(I'm thinking of approaching it this way because, for `impl`s, we can have more than one non-defining declaration in a file (due to `match_first` blocks and `extend impl` declarations), so we'd be able to reach these checks for `impl`s when neither declaration is a definition.)